src/charm.py

```diff
@@ -113,372 +126,145 @@ def build_charm_state(self) -> CharmState:
-        if secret_id:
-            try:
-                secret = self.model.get_secret(id=secret_id)
-                logging.debug(secret.get_content().get("secret-signing-key"))
-                return secret.get_content().get("secret-signing-key")
-            except (ops.model.SecretNotFoundError, ValueError, TypeError) as exc:
-                logger.exception("Failed to get secret id %s: %s", secret_id, str(exc))
-                del peer_relation[0].data[self.app]["secret-signing-id"]
-        return None
+        check_mjolnir = False
+        try:
+            service = container.get_services(synapse.MJOLNIR_SERVICE_NAME)
+            if service:
+                check_mjolnir = True
+        except ops.ModelError:
+            # ModelError is raised if service not found
+            logger.info("mjolnir not found, skipping")
+        if check_mjolnir and not self._is_service_running(container, synapse.MJOLNIR_SERVICE_NAME):
+            self.unit.status = ops.MaintenanceStatus("Waiting for Mjolnir")
+            self._install_callback_script(synapse.MJOLNIR_SERVICE_NAME)
+            container.start(f"{synapse.MJOLNIR_SERVICE_NAME}-charm-callback")
+            return
 
-    @inject_charm_state
-    def _on_leader_elected(self, _: ops.HookEvent, charm_state: CharmState) -> None:
-        """Handle Synapse leader elected event.
+        self.unit.status = ops.ActiveStatus()
 
-        This event handler will reconcile Synapse configuration after the following
-        scenarios:
-        - When the charm is deployed so the leader will be the main unit.
-        - When the leader, for any reason, has changed so the leader unit will be the main.
-        Once the peer data (main_unit_id) is changed, other units will emit reconcile and be
-        properly configured.
+    def _install_callback_script(self, service_name: str) -> None:
+        """Install platform startup callback script for noticing the charm on start.
 
         Args:
-            charm_state: The charm state.
+            service_name: service name to check status.
         """
-        # assuming that this event will be fired only at the setup phase
-        # check if main is already set if not, this unit will be the main
-        if not self.unit.is_leader():
-            return
-        logging.debug(
-            "_on_leader_elected received, main_unit is %s and will be set to %s",
-            self.get_main_unit(),
-            self.unit.name,
+        callback_script_path = pathlib.Path(
+            CALLBACK_SCRIPT_FILENAME.format(service_name=service_name)
         )
-        self.set_main_unit(self.unit.name)
-        logger.debug("_on_leader_elected emitting reconcile")
-        self.reconcile(charm_state)
-
-    @inject_charm_state
-    def _on_relation_changed(self, _: ops.HookEvent, charm_state: CharmState) -> None:
-        """Handle Synapse peer relation changed event.
+        script = textwrap.dedent(
+            f"""\
+            while :; do
```

Comment:
super-minor: while this is totally valid, it could be more readable for mere mortals as:


```suggestion
            while true; do
```